arbnode/espresso_batcher_addr_monitor.go

```diff
@@ -0,0 +1,438 @@
+	seqInboxInterface *bridgegen.SequencerInbox
+	deployAt          uint64
+}
+
+func NewBatcherAddrMonitor(
+	initAddresses []common.Address,
+	db ethdb.Database,
+	l1Reader *headerreader.HeaderReader,
+	seqInboxAddr common.Address,
+	deployAt uint64,
+) *BatcherAddrMonitor {
+	seqInboxInterface, err := bridgegen.NewSequencerInbox(seqInboxAddr, l1Reader.Client())
+	if err != nil {
+		panic(err)
+	}
+	return &BatcherAddrMonitor{
+		initAddresses:     initAddresses,
+		db:                db,
+		l1Reader:          l1Reader,
+		seqInboxAddr:      seqInboxAddr,
+		seqInboxInterface: seqInboxInterface,
+		deployAt:          deployAt,
+		parentHeight:      deployAt,
+	}
+}
+
+func (b *BatcherAddrMonitor) AddBatchPosterSetEvents(events []BatcherAddrEvent) error {
+	if len(events) == 0 {
+		return nil
+	}
+	b.events = append(b.events, events...)
+	// Sort events by l1Height to ensure correct processing order.
+	// Since BatcherAddr events are infrequent, the performance impact of sorting is negligible.
+	sort.Slice(b.events, func(i, j int) bool {
+		return b.events[i].L1Height < b.events[j].L1Height
+	})
+	b.lastEventL1Height = b.events[len(b.events)-1].L1Height
+	b.cached = false
+	return b.Store()
+}
+
+func (b *BatcherAddrMonitor) GetValidAddresses(l1 uint64) []common.Address {
+	if l1 > b.l1Height {
+		// If the target L1 height is greater than the latest known L1 height,
+		// return an empty slice. The caller should wait until the monitor has
+		// observed at least this L1 height before calling this function.
+		return []common.Address{}
+	}
+
+	if len(b.events) == 0 || b.events[0].L1Height > l1 {
+		return b.initAddresses
+	}
+
+	// If the target L1 height is within the latest cached window, return the cached result.
+	// In a practical scenario, this is the most common case. Here means that during the time
+	// from `lastEventL1Height` to `l1Height`, the `events` are not changed. It is not needed to
+	// calculate valid batcher addresses.
+	latestCachedWindow := l1 >= b.lastEventL1Height && l1 <= b.l1Height
+	if b.cached && latestCachedWindow {
+		return b.cachedAddresses
+	}
+
+	result := map[common.Address]bool{}
+	for _, addr := range b.initAddresses {
+		result[addr] = true
+	}
+
+	for _, event := range b.events {
+		if event.L1Height > l1 {
+			break
+		}
+
+		result[event.Addr] = event.IsBatcher
+	}
+
+	var validAddrs []common.Address
+	for addr, isBatcher := range result {
+		if isBatcher {
+			validAddrs = append(validAddrs, addr)
+		}
+	}
+
+	if latestCachedWindow {
+		b.cached = true
+		b.cachedAddresses = validAddrs
+	}
+
+	return validAddrs
+}
+
+func (b *BatcherAddrMonitor) SetParentHeight(height uint64) {
+	b.parentHeight = height
+}
+
+func (b *BatcherAddrMonitor) SetL1Height(height uint64) {
+	b.l1Height = height
+}
+
+func (b *BatcherAddrMonitor) GetConfirmedParentHeight() uint64 {
+	return b.parentHeight
+}
+
+func (b *BatcherAddrMonitor) LookupEvents(ctx context.Context, fromBlock, toBlock uint64) ([]BatcherAddrEvent, error) {
```

Comment:
very generic name: change name to LookupOwnerFunctionCalledEvent. Add some comments here

Reply:
I see the confusion here. Here the `events` means the BatcherAddrEvent I defined in this file. Will rename it